nova/scheduler/filters/external_scheduler_filter.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2024 SAP SE or an SAP affiliate company.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+
+import requests
+
+import nova.conf
+from nova.scheduler import filters
+from oslo_log import log as logging
+
+LOG = logging.getLogger(__name__)
+CONF = nova.conf.CONF
+
+
+class ExternalSchedulerFilter(filters.BaseHostFilter):
+    """Filter using an out of tree scheduler via a REST API."""
+
+    RUN_ON_REBUILD = False
+
+    def __init__(self):
+        super().__init__()
+        self.api_url = CONF.filter_scheduler.external_scheduler_api_url
+
+    def filter_all(self, filter_obj_list, spec_obj):
+        """Yield objects that pass the filter."""
+        if not self.api_url:
+            LOG.debug(
+                "external_scheduler_api_url not configured. skipping filter"
+            )
+            return filter_obj_list
+
+        hosts_data = []
+        for obj in filter_obj_list:
+            hosts_data.append({
+                "host_name": obj.host,
+                "host_status": obj.status,
+                "vcpus_total": obj.vcpus_total,
+                "vcpus_used": obj.vcpus_used,
+                "memory_mb": obj.memory_mb,
+                "memory_mb_used": obj.memory_mb_used
+            })
+
+        try:
+            spec_dict = vars(spec_obj)
```

Comment:
Vielleicht lieber: `spec_obj.obj_to_primitive()`

Reply:
Thanks!
I did it explicitly now for vcpus, memory. 
Deriving this from the object results in some bad json.

Reply:
The problem is, now you have only a subset of the fields, and no version-negotiation.
You can still pick out just the fields you want on the receiver side on from oslo_versionedobjects json.
I agree, it is harder to do that from there, but I am a big fan of working with the tools/libraries as they are intended.

Reply:
As to the work on the other side: Since [zed](https://github.com/openstack/oslo.versionedobjects/commit/92e7ac7108d03137468f1918497ffa71329d0347), there is a class-method `to_json_schema`, which will
drop you the schema, which in turn could be used to generate the full object.

Here would be one of such tools: https://github.com/omissis/go-jsonschema

I agree, doing all that is a bit much to start with, but using here `spec_obj.obj_to_primitive()` opens the path for that, and the usage of other serializers (protobuf?).  Parsing "manually" the admittedly ugly json for that is in my mind a small price to pay for that.

Reply:
Not that I will block this change because of that.